Update the GraphQL Tool Guide with SDL Schema Generation (#6063)
ballerina-platform/ballerina-dev-website: swan-lake/learn-the-platform/ballerina-tooling/graphql-tool.md

---
layout: ballerina-graphql-support-left-nav-pages-swanlake
title: GraphQL tool

> description: Check out how the Ballerina GraphQL tool makes it easy for you to generate a client in Ballerina using a given GraphQL config file and a GraphQL schema using a given Ballerina GraphQL service.

Author: This changes the idea. There are two tasks that can be done using the GraphQL tool,
1. generate a client in Ballerina using a given GraphQL config file.
2. generate a GraphQL schema using a given Ballerina GraphQL service.
```suggestion
description: Check out how the Ballerina GraphQL tool makes it easy for you to generate a client in Ballerina using a given GraphQL config file and generate a GraphQL schema using a given Ballerina GraphQL service.
```

> description: Check out how the Ballerina GraphQL tool makes it easy for you to generate a client in Ballerina using a given GraphQL config file and a GraphQL schema using a given Ballerina GraphQL service.

Reviewer: +1 for this.

keywords: ballerina, programming language, graphql, sdl, schema definition language
permalink: /learn/graphql-tool/
active: graphql-tool

> intro: The Ballerina GraphQL tool will make it easy for you to start the development of a GraphQL client. It generates the client skeletons in Ballerina for a given GraphQL schema and a GraphQL document. In addition, the GraphQL tool supports generating the GraphQL schema specified by the Schema Definition Language for a given Ballerina GraphQL service.

Reviewer 2: ```suggestion
intro: The Ballerina GraphQL tool will make it easy for you to start the development of a GraphQL client. It generates the client skeletons in Ballerina for a given GraphQL schema and a GraphQL document. In addition, the GraphQL tool supports generating the GraphQL schema for a given Ballerina GraphQL service and write it to a file in GraphQL schema definition language(SDL).
```

redirect_from:
  - /learn/ballerina-graphql-support
  - /learn/ballerina-graphql-support/
  - /learn/graphql-tool
  - /learn/graphql-tool/
--- 

The Ballerina GraphQL tooling support provides the following capabilities.

1. Generating a Ballerina client from a given GraphQL config file configured with a GraphQL schema specified by Schema Definition Language and a GraphQL document.

2. Generating a Ballerina client from a given GraphQL config file configured with a GraphQL schema specified by Schema Definition Language and multiple GraphQL documents.

3. Generating multiple Ballerina modules from a given GraphQL config file configured with multiple GraphQL projects.

> 4. Generate the GraphQL schema specified by the Schema Definition Language for a given Ballerina GraphQL service(s).

Reviewer 2: This doesn't make sense.

```suggestion
4. Generate the GraphQL schema for a given Ballerina GraphQL service(s) and write the schema(s) to a file/files using GraphQL schema definition language.
```

> >**Note:** You are required to give a query name for each query. This name is used as the corresponding remote operation name in the generated client. 

Reviewer: ```suggestion
>**Note:** You are required to give a name for each query. This name is used as the corresponding remote operation name in the generated client. 
```

2. `utils.bal` - contains all the utility methods related to the client stub
3. `types.bal`- contains all the Ballerina data types extracted from the GraphQL schema (SDL)

You can use the generated modules to work with multiple GraphQL APIs. Use the client that is generated within each module to execute the queries/mutations on the relevant GraphQL API. Also, use the clients that are generated **within** the same Ballerina package by importing the generated modules. For more information, see [Import a module from the same package](/learn/organize-ballerina-code/#import-a-module-from-the-same-package).

> ## Schema Generation

Reviewer: ```suggestion
## Schema generation
```

> The GraphQL tool supports generating the GraphQL schema specified by the Schema Definition Language for a given Ballerina GraphQL service. You will be able to export the generated schema to the given location. The tool can be used in the following ways to generate the schema.

Reviewer 2: The schema is not specified by the Schema Definition Language. In Ballerina, schema is defined using the Ballerina service. Furthermore, SDL is a generic term, not specific to GraphQL. Shall we reword this accordingly?

```suggestion
The GraphQL tool supports generating the GraphQL schema specified by the Schema Definition Language for a given Ballerina GraphQL service. You will be able to export the generated schema to the given location. The tool can be used in the following ways to generate the schema.
```

### Generate GraphQL schemas for all services

If your Ballerina file includes multiple GraphQL services, this command generates the GraphQL schema for each service in the Ballerina file.

```
$ bal graphql [-i | --input] <ballerina-service-file-path> [(-o | --output) <output-location>]
```

For example,

```
$ bal graphql -i service.bal -o ./schema
```

The `input` parameter specifies the path of the ballerina service file (e.g., `service.bal`) and is mandatory.

>**Info:** For more information on the command, see [Ballerina to GraphQL](/learn/cli-documentation/graphql/#ballerina-to-graphql).

> ### Generate a schema for a specific service.

Reviewer: ```suggestion
### Generate a schema for a specific service
```
